### PR TITLE
Skip reading Parquet pages using Column Indexes feature of Parquet 1.…

### PR DESCRIPTION
**Pull request by shangxinli**
Parquet column index is a feature that can speed up reading of Parquet files by skipping unnecessary readings of parquet pages.  After upgrading Pesto to Parquet 1.11.x, it is time to integrate this feature into Presto. 

Test Plan - 1) Added unit tests. 2) Tested in staging with test tables. 

```
== RELEASE NOTES ==

General Changes
* When there are page statistics in the data, Presto read the statistics and use it to filter out row groups first. And then use it to filter out pages that are not in the range of the filter. 
* The column readers are synchronized across columns based on row ranges. The synchronization makes the columns that are not with filters can also be effective skip pages.   

Hive Changes
* Add a session-based flag 'hive.readColumnIndexFilter'. When it is false(default), this feature is disabled. When it is on and there is page statistics in the Parquet file, this feature is invoked. 
* The ParquetDataSource is added new methods to read column index and offset index. In case there is no statistics, null is returned. 
```

### CODE REVIEW

File: presto-hive/src/main/java/com/facebook/presto/hive/HiveSessionProperties.java [hunk 1]
```diff
@@ -1023,6 +1029,7 @@ public static boolean isFileRenamingEnabled(ConnectorSession session)
         return session.getProperty(FILE_RENAMING_ENABLED, Boolean.class);
     }
 
+<<<<<<< HEAD
```
[the remark this answers is not in the document]
Comment (shangxinli) [reply presto-hive/src/main/java/com/facebook/presto/hive/HiveSessionProperties.java:1032]: Yeah, not ready to review it. I will tag you when it is ready.

**Comment by shangxinli** [comment]
@beinan It is ready to review the code. Thanks for your time in advance!

File: presto-parquet/src/main/java/com/facebook/presto/parquet/predicate/TupleDomainParquetPredicate.java [hunk 2]
```diff
@@ -352,4 +472,354 @@ private static void failWithCorruptionException(boolean failOnCorruptedParquetSt
...
+        }
+
+        public List<Long> getMaxValuesAsLong(Type type, ColumnIndex columnIndex, String column)
+        {
+            int pageCount = columnIndex.getMaxValues().size();
+            List<ByteBuffer> maxValues = columnIndex.getMaxValues();
+            List<Long> maxs = new ArrayList<>();
+            if (TINYINT.equals(type) || SMALLINT.equals(type) || INTEGER.equals(type)) {
+                for (int i = 0; i < pageCount; i++) {
+                    int maxValue = converter.convert(maxValues.get(i), column);
```
Comment (beinan) [inline presto-parquet/src/main/java/com/facebook/presto/parquet/predicate/TupleDomainParquetPredicate.java:716]: inline maxValue?

File: presto-parquet/src/main/java/com/facebook/presto/parquet/predicate/TupleDomainParquetPredicate.java [hunk 3]
```diff
@@ -352,4 +472,354 @@ private static void failWithCorruptionException(boolean failOnCorruptedParquetSt
...
+                    maxs.add((long) maxValue);
+                }
+            }
+            else if (BIGINT.equals(type)) {
+                for (int i = 0; i < pageCount; i++) {
+                    long maxValue = converter.convert(maxValues.get(i), column);
+                    maxs.add((long) maxValue);
```
Comment (beinan) [inline presto-parquet/src/main/java/com/facebook/presto/parquet/predicate/TupleDomainParquetPredicate.java:723]: same as above

File: presto-parquet/src/main/java/com/facebook/presto/parquet/predicate/TupleDomainParquetPredicate.java [hunk 4]
```diff
@@ -352,4 +472,354 @@ private static void failWithCorruptionException(boolean failOnCorruptedParquetSt
...
+                }
+            }
+            else if (REAL.equals(type)) {
+                for (int i = 0; i < pageCount; i++) {
+                    float maxValue = floatToRawIntBits(converter.convert(maxValues.get(i), column));
```
Comment (beinan) [inline presto-parquet/src/main/java/com/facebook/presto/parquet/predicate/TupleDomainParquetPredicate.java:728]: same as above

File: presto-parquet/src/main/java/com/facebook/presto/parquet/predicate/TupleDomainParquetPredicate.java [hunk 5]
```diff
@@ -352,4 +472,354 @@ private static void failWithCorruptionException(boolean failOnCorruptedParquetSt
...
+        }
+
+        public List<Double> getMinValuesAsDouble(Type type, ColumnIndex columnIndex, String column)
+        {
+            int pageCount = columnIndex.getMinValues().size();
+            List<ByteBuffer> minValues = columnIndex.getMinValues();
+            List<Double> mins = new ArrayList<>();
+            if (DOUBLE.equals(type)) {
+                for (int i = 0; i < pageCount; i++) {
+                    double minValue = converter.convert(minValues.get(i), column);
```
Comment (beinan) [inline presto-parquet/src/main/java/com/facebook/presto/parquet/predicate/TupleDomainParquetPredicate.java:743]: inline minValue?

File: presto-parquet/src/main/java/com/facebook/presto/parquet/predicate/TupleDomainParquetPredicate.java [hunk 6]
```diff
@@ -352,4 +472,354 @@ private static void failWithCorruptionException(boolean failOnCorruptedParquetSt
...
+                    mins.add(minValue);
+                }
+            }
+            //TODO: Else
+            return mins;
+        }
+
+        public List<Double> getMaxValuesAsDouble(Type type, ColumnIndex columnIndex, String column)
```
Comment (beinan) [inline presto-parquet/src/main/java/com/facebook/presto/parquet/predicate/TupleDomainParquetPredicate.java:751]: Any chance to make these getMin/MaxValues methods more generic?

File: presto-hive/src/main/java/com/facebook/presto/hive/parquet/ParquetPageSourceFactory.java [hunk 7]
```diff
@@ -337,6 +353,31 @@ else if (getParquetType(type, fileSchema, useParquetColumnNames, column, tableNa
...
+
+        if (!hasColumnIndex) {
+            return null;
+        }
+
+        Set<ColumnPath> paths = new HashSet<>();
+        for (List<String> path : descriptorsByPath.keySet()) {
+            paths.add(ColumnPath.get(path.toArray(new String[0])));
+        }
+        return ColumnIndexStoreImpl.create(dataSource, blockMetadata, paths);
```
[the remark this answers is not in the document]
Comment (beinan) [reply presto-hive/src/main/java/com/facebook/presto/hive/parquet/ParquetPageSourceFactory.java:378]: better?  your call

**Review by beinan — COMMENTED** [review]
Thanks a lot for this great contribution @shangxinli !  Just left a couple of comments, but I have not finished the review yet,  I will continue the review in the following days.

File: presto-hive/src/main/java/com/facebook/presto/hive/HiveClientConfig.java [hunk 8]
```diff
@@ -1682,4 +1684,17 @@ public int getMaterializedViewMissingPartitionsThreshold()
     {
         return this.materializedViewMissingPartitionsThreshold;
     }
+
+    @Config("hive.readColumnIndexFilter")
```
[the remark this answers is not in the document]
Comment (chliang71) [reply presto-hive/src/main/java/com/facebook/presto/hive/HiveClientConfig.java:1688]: apart from the name, I think the naming convention here should be something like
hive.read-column-index-filter

File: presto-parquet/src/main/java/com/facebook/presto/parquet/reader/PageReader.java [hunk 9]
```diff
@@ -112,4 +134,27 @@ public DictionaryPage readDictionaryPage()
...
+    /**
+     * Cast value to a an int, or throw an exception
+     * if there is an overflow.
+     *
+     * @param value a long to be casted to an int
+     * @return an int that is == to value
+     * @throws IllegalArgumentException if value can't be casted to an int
+     * @deprecated replaced by {@link java.lang.Math#toIntExact(long)}
+     */
+    public static int checkedCast(long value)
```
Comment (chliang71) [inline presto-parquet/src/main/java/com/facebook/presto/parquet/reader/PageReader.java:147]: curious why not just use java.lang.Math.toIntExact here?

File: presto-hive/src/main/java/com/facebook/presto/hive/HiveClientConfig.java (hunk 8 above)
[the remark this answers is not in the document]
Comment (shangxinli) [reply presto-hive/src/main/java/com/facebook/presto/hive/HiveClientConfig.java:1688]: I changed it to hive.parquet-use-column-index-filter. Let me know if this makes sense.

File: presto-hive/src/main/java/com/facebook/presto/hive/HiveSessionProperties.java [hunk 10]
```diff
@@ -129,6 +129,7 @@
     public static final String VERBOSE_RUNTIME_STATS_ENABLED = "verbose_runtime_stats_enabled";
     private static final String DWRF_WRITER_STRIPE_CACHE_ENABLED = "dwrf_writer_stripe_cache_enabled";
     private static final String DWRF_WRITER_STRIPE_CACHE_SIZE = "dwrf_writer_stripe_cache_size";
+    public static final String READ_COLUMN_INDEX_FILTER = "read_column_index_filter";
```
[the remark this answers is not in the document]
Comment (shangxinli) [reply presto-hive/src/main/java/com/facebook/presto/hive/HiveSessionProperties.java:132]: I changed it to use_column_index_filter. Let me know if it makes sense.

File: presto-hive/src/main/java/com/facebook/presto/hive/parquet/HdfsParquetDataSource.java [hunk 11]
```diff
@@ -67,6 +73,28 @@ protected void readInternal(long position, byte[] buffer, int bufferOffset, int
         }
     }
 
+    @Override
+    public ColumnIndex readColumnIndex(ColumnChunkMetaData column) throws IOException
+    {
+        IndexReference ref = column.getColumnIndexReference();
+        if (ref == null) {
```
[the remark this answers is not in the document]
Comment (shangxinli) [reply presto-hive/src/main/java/com/facebook/presto/hive/parquet/HdfsParquetDataSource.java:80]: let's use indexRef

File: presto-hive/src/main/java/com/facebook/presto/hive/parquet/HdfsParquetDataSource.java [hunk 12]
```diff
@@ -67,6 +73,28 @@ protected void readInternal(long position, byte[] buffer, int bufferOffset, int
...
+            return null;
+        }
+        inputStream.seek(ref.getOffset());
+        return ParquetMetadataConverter.fromParquetColumnIndex(column.getPrimitiveType(), Util.readColumnIndex(inputStream));
+    }
+
+    @Override
+    public OffsetIndex readOffsetIndex(ColumnChunkMetaData column) throws IOException
+    {
+        IndexReference ref = column.getOffsetIndexReference();
```
[the remark this answers is not in the document]
Comment (shangxinli) [reply presto-hive/src/main/java/com/facebook/presto/hive/parquet/HdfsParquetDataSource.java:90]: Got it.

**Comment by zhenxiao** [comment]
hi @shangxinli any updates on this PR? could you please rebase and squash all commits into one? This is a nice feature. I'd like to take a look

**Comment by shangxinli** [comment]
@zhenxiao Thanks for looking into it. I addressed the earlier feedback from @beinan and am waiting for him to have another look.  Yeah, I can squash and commit.

**Comment by shangxinli** [comment]
For some reasons, my branch is not able to rebase/squash because of [commit](https://github.com/prestodb/presto/pull/16963/commits/4476595f0dde4c133e6061ec7f1611fca65e5ae2). I can create a new branch and resumit a new PR. Let me know if you want me to do so @zhenxiao @beinan

**Comment by zhenxiao** [comment]
thank you, @shangxinli 
feel free to squash the commits. Or, submit new PRs are good, too :)

**Pull request closed without merging**